scipy/_lib/tests/test_public_api.py

```diff
@@ -0,0 +1,324 @@
+import pkgutil
+import types
+import importlib
+import warnings
+
+import scipy
+
+
+def check_dir(module, module_name=None):
+    """Returns a mapping of all objects with the wrong __module__ attribute."""
+    if module_name is None:
+        module_name = module.__name__
+    results = {}
+    for name in dir(module):
+        item = getattr(module, name)
+        if (hasattr(item, '__module__') and hasattr(item, '__name__')
+                and item.__module__ != module_name):
+            results[name] = item.__module__ + '.' + item.__name__
+    return results
+
+
+def test_scipy_namespace():
+    # None of these objects are publicly documented to be part of the main
+    # SciPy namespace (some are useful though, others need to be cleaned up)
+    undocumented = {
+        'LowLevelCallable': 'scipy._lib._ccallback.LowLevelCallable'
+    }
+    # We override dir to not show these members
+    allowlist = undocumented
+    bad_results = check_dir(scipy)
+    # pytest gives better error messages with the builtin assert than with
+    # assert_equal
+    assert bad_results == allowlist
+
+
+def test_dir_testing():
+    """Assert that output of dir has only one "testing/tester"
+    attribute without duplicate"""
+    assert len(dir(scipy)) == len(set(dir(scipy)))
+
+
+# Historically SciPy has not used leading underscores for private submodules
+# much.  This has resulted in lots of things that look like public modules
+# (i.e. things that can be imported as `import scipy.somesubmodule.somefile`),
+# but were never intended to be public.  The PUBLIC_MODULES list contains
+# modules that are either public because they were meant to be, or because they
+# contain public functions/objects that aren't present in any other namespace
+# for whatever reason and therefore should be treated as public.
+#
+# The PRIVATE_BUT_PRESENT_MODULES list contains modules that look public (lack
+# of underscores) but should not be used.  For many of those modules the
+# current status is fine.  For others it may make sense to work on making them
+# private, to clean up our public API and avoid confusion.
+
+PUBLIC_MODULES = ["scipy." + s for s in [
+    "cluster",
+    "cluster.vq",
+    "cluster.hierarchy",
+    "constants",
+    "fft",
+    "fftpack",
+    "integrate",
+    "interpolate",
+    "io",
+    "io.arff",
+    "io.matlab",
+    "io.wavfile",
+    "linalg",
+    "linalg.blas",
+    "linalg.cython_blas",
+    "linalg.lapack",
+    "linalg.cython_lapack",
+    "linalg.interpolative",
+    "misc",
+    "ndimage",
+    "odr",
+    "optimize",
+    "signal",
+    "signal.windows",
+    "sparse",
+    "sparse.linalg",
+    "sparse.csgraph",
+    "spatial",
+    "spatial.distance",
+    "spatial.transform",
+    "special",
+    "stats",
+    "stats.distributions",
+    "stats.mstats",
+    "stats.qmc",
+]]
+
+
+PRIVATE_BUT_PRESENT_MODULES = [
```

Comment:
Are we tracking the "privatization" of this somewhere? IMO we should, and add a comment here.

It seems all boxes in https://github.com/scipy/scipy/issues/14360 are checked, but AFAIU, `PRIVATE_BUT_PRESENT_MODULES` should ideally be empty and is a stopgap for now?

Reply:
> IMO we should, and add a comment here.

Done. Thanks

> `PRIVATE_BUT_PRESENT_MODULES` should ideally be empty and is a stopgap for now?

Yes. They can be removed once we clear private modules support

Reply:
> Done. Thanks

Cool thanks!

> These private modules support will be removed in SciPy v2.0.0

Could you open an issue for this (assuming there isn't one yet), and mention that in the comment here as well?

Reply:
Let's not open an issue specifically for this, because it won't be actionable for a long time. At most I'd like one tracking issue for things to go into `v2.0.0`. That could grow over time with items we really want or need to do, and cannot have in `1.xx` releases.

Reply:
> Let's not open an issue specifically for this

OK, apologies

> because it won't be actionable for a long time.

That part I don't understand though. scipy isn't strictly following semver, so if the decision is made to make private modules actually private, why not follow through on that slowly but steadily? And why not have a tracking issue (e.g. to discuss timelines), regardless how long it takes...?

Certainly, I defer to Ralf on this - it was just how I'd have handled it.

Reply:
With deprecations it's always a cost-benefit thing. We have a minimum of two minor releases for removing of deprecated code, and no maximum. In this case there's a very large amount of changes, and the cost of keeping the compatibility shims is not all that high. It'd be a nice cleanup, but I'm already happy we are finally doing this anyway.

I will open an issue now.

Reply:
https://github.com/scipy/scipy/wiki/Ideas-for-changes-&-cleanups-in-SciPy-2.0